File: transpiler/javatests/com/google/j2cl/readable/java/nullability/Nullability.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nullability;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import javax.annotation.Nonnull;
import jsinterop.annotations.JsConstructor;
import jsinterop.annotations.JsFunction;
import jsinterop.annotations.JsMethod;
import jsinterop.annotations.JsNonNull;
import org.jspecify.annotations.Nullable;

public class Nullability {
  @Nonnull private String f1 = "Hello";
  private String f2 = null;
  @Nullable private String f3 = null;
  private @Nonnull List<@JsNonNull String> f4 = new ArrayList<>();
  private List<@JsNonNull String> f5 = new ArrayList<>();
  private @Nonnull List<String> f6 = new ArrayList<>();
  private List<String> f7 = null;
  private @JsNonNull String @JsNonNull [] f8 = {};
  // Nonnullable array of nullable strings.
  private String @JsNonNull [] f9 = {};
  // Nullable array of non-nullable strings.
  private @JsNonNull String[] f10 = {};
  // Conversion from generic type parameter.
  private List<String> f12 = new ArrayList<>();
  @Nonnull private Object f13;
  private Object f14;

  @JsConstructor
  public Nullability(@Nonnull String a) {
    f13 = new Object();
  }

  @Nonnull
  public String m1(@Nonnull String a, @Nonnull List<@JsNonNull Double> b, String c) {
    return "";
  }

  public String m2(String a, @Nonnull List<@Nullable Double> b) {
    return "";
  }

  @JsMethod
  public String m3(@Nonnull String a, @JsNonNull String... args) {
    return null;
  }

  public void m4(@Nonnull MyFunction f) {}

  static class Foo<T> {
    void bar(T t) {}

    @Nullable
    T baz() {
      return null;
    }

    @Override
    public String toString() {
      return "Foo";
    }
  }

  @JsFunction
  interface MyFunction {
    @Nonnull
    String x(@Nonnull String a);
  }

  public interface StringList extends List<String> {}

  // Should implement Comparator<string>
  public static class StringComparator implements Comparator<@JsNonNull String> {
    @Override
    public int compare(@Nonnull String a, @Nonnull String b) {
      return 0;
    }
  }

  // Should implement Comparator<?string>
  public static class NullableStringComparator implements Comparator<@Nullable String> {
    @Override
    public int compare(@Nullable String a, @Nullable String b) {
      return 0;
    }
  }

  interface NullableTemplatedReturn<T> {
    @Nullable
    T foo();
  }

  public static class NullableTemplatedReturnOverride implements NullableTemplatedReturn<String> {
    @Nullable
    @Override
    public String foo() {
      return "foo";
    }
  }

  static class ParameterizedDefaultNullability<N> {
    @Nullable N getNullable() {
      return null;
    }

    @JsNonNull
    N getNonNullable() {
      throw new RuntimeException();
    }

    N getDefaultNullability() {
      return null;
    }
  }

  static class ParameterizedNullable<N extends @Nullable Object> {
    @Nullable N getNullable() {
      return null;
    }

    @JsNonNull
    N getNonNullable() {
      throw new RuntimeException();
    }

    N getDefaultNullability() {
      return null;
    }
  }

  static class ParameterizedNonNullable<N extends @JsNonNull Object> {
    @Nullable N getNullable() {
      return null;
    }

    @JsNonNull
    N getNonNullable() {
      throw new RuntimeException();
    }

    N getDefaultNullability() {
      return null;
    }
  }

  static class NonNullableInsideNullable<T extends @Nullable Object> {
    void nonNullableTest(ParameterizedNonNullable<T> nonNullable) {}
  }

  static <T extends String> void testDefaultNullabilityBounds(
      ParameterizedDefaultNullability<T> defaultNullability) {
    defaultNullability.getNonNullable().length();
    defaultNullability.getNullable().length();
    defaultNullability.getDefaultNullability().length();
  }

  static <T extends String> void testDefaultNullabilityBounds(ParameterizedNullable<T> nullable) {
    nullable.getNonNullable().length();
    nullable.getNullable().length();
    nullable.getDefaultNullability().length();
  }

  static <T extends @Nullable String> void testNullableBounds(
      ParameterizedDefaultNullability<T> defaultNullability) {
    defaultNullability.getNonNullable().length();
    defaultNullability.getNullable().length();
    defaultNullability.getDefaultNullability().length();
  }

  static <T extends @Nullable String> void testNullableBounds(ParameterizedNullable<T> nullable) {
    nullable.getNonNullable().length();
    nullable.getNullable().length();
    nullable.getDefaultNullability().length();
  }

  static <T extends @JsNonNull String> void testNonNullableBounds(
      ParameterizedDefaultNullability<T> defaultNullability) {
    defaultNullability.getNonNullable().length();
    defaultNullability.getNullable().length();
    defaultNullability.getDefaultNullability().length();
  }

  static <T extends @JsNonNull String> void testNonNullableBounds(
      ParameterizedNullable<T> nullable) {
    nullable.getNonNullable().length();
    nullable.getNullable().length();
    nullable.getDefaultNullability().length();
  }

  static <T extends @JsNonNull String> void testNonNullableBounds(
      ParameterizedNonNullable<T> nonNullable) {
    nonNullable.getNonNullable().length();
    nonNullable.getNullable().length();
    nonNullable.getDefaultNullability().length();
  }

  static void testDefaultNullabilityWildcards(
      ParameterizedDefaultNullability<? extends String> nonNullable) {
    nonNullable.getNonNullable().length();
    nonNullable.getNullable().length();
    nonNullable.getDefaultNullability().length();
  }

  static void testDefaultNullabilityWildcards(ParameterizedNullable<? extends String> nonNullable) {
    nonNullable.getNonNullable().length();
    nonNullable.getNullable().length();
    nonNullable.getDefaultNullability().length();
  }

  static void testDefaultNullabilityWildcards(
      ParameterizedNonNullable<? extends String> nonNullable) {
    nonNullable.getNonNullable().length();
    nonNullable.getNullable().length();
    nonNullable.getDefaultNullability().length();
  }

  static class Recursive<T extends @JsNonNull Recursive<T> & Marker> {}

  static class RecursiveNullable<T extends RecursiveNullable<T> & Marker> {}

  static class RecursiveChild extends Recursive<RecursiveChild> implements Marker {}

  static class RecursiveNullableChild extends RecursiveNullable<RecursiveNullableChild>
      implements Marker {}

  static class RecursiveParam<T extends Recursive<T> & Marker> {}

  static class RecursiveNullableParam<T extends RecursiveNullable<T> & Marker> {}

  static <T extends Recursive<T> & Marker> void testRecursive() {
    RecursiveParam<T> generic = new RecursiveParam<>();
    RecursiveParam<RecursiveChild> parametrized = new RecursiveParam<>();
  }

  static <T extends RecursiveNullable<T> & Marker> void testRecursiveNullable() {
    RecursiveNullableParam<T> generic = new RecursiveNullableParam<>();
    RecursiveNullableParam<RecursiveNullableChild> parametrized = new RecursiveNullableParam<>();
  }
}

interface Marker {}

File: transpiler/javatests/com/google/j2cl/readable/kotlin/nullability/Nullability.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package nullability

import jsinterop.annotations.JsConstructor
import jsinterop.annotations.JsMethod
import jsinterop.annotations.JsOptional

class Nullability @JsConstructor constructor(a: String) {
  private val f1: String = "Hello"
  private val f2: String? = null

  private val f4: List<String> = ArrayList()
  private var f5: List<String>? = ArrayList()
  private val f6: List<String?> = ArrayList()
  private val f7: List<String?>? = null

  private val f8: Array<String> = arrayOf()
  // Nonnullable array of nullable strings.
  private val f9: Array<String?> = arrayOf()
  // Nullable array of non-nullable strings.
  private val f10: Array<String>? = arrayOf()
  // Conversion from generic type parameter.
  private val f12: List<String?>? = ArrayList()

  private val f13: Any
  private val f14: Any? = null

  init {
    f13 = Any()
  }

  fun m1(a: String, b: List<Double>, c: String?): String {
    return ""
  }

  fun m2(a: String?, b: List<Double?>): String? {
    return ""
  }

  fun m3(a: String?, vararg args: String): String? {
    return null
  }

  fun m4(f: MyFunction) {}

  @JsMethod fun m5(a: String, @JsOptional jsOptional: String?) {}

  fun testPrimitive() {
    val a: Boolean = true
    val b: Boolean? = true
    val c: Byte = 0
    val d: Byte? = null
    val e: Char = 'a'
    val f: Char? = null
    val g: Double = 0.0
    val h: Double? = null
    val i: Float = 0.0f
    val j: Float? = null
    val k: Int = 0
    val l: Int? = null
    val m: Long = 0
    val n: Long? = null
    val o: Short = 0
    val p: Short? = null
    val q: Unit? = null
  }
}

class Foo<T> {
  fun bar(t: T) {}

  fun baz(): T? {
    return null
  }

  @JsMethod fun foo(a: String, @JsOptional jsOptional: T?) {}

  override fun toString(): String {
    return "Foo"
  }
}

interface MyFunction {
  fun x(a: String): String
}

interface StringList : List<String?>

// Should implement Comparator<string>
class StringComparator : Comparator<String> {
  override fun compare(a: String, b: String): Int {
    return 0
  }
}

// Should implement Comparator<?string>
class NullableStringComparator : Comparator<String?> {
  override fun compare(a: String?, b: String?): Int {
    return 0
  }
}

internal interface NullableTemplatedReturn<T> {
  fun foo(): T?
}

class NullableTemplatedReturnOverride : NullableTemplatedReturn<String?> {
  override fun foo(): String? {
    return "foo"
  }
}

internal class ParameterizedNonNullable<N : Any> {
  fun getNullable(): N? = null

  fun getDefaultNullability(): N = throw RuntimeException()
}

internal class ParameterizedNullable<N> {
  fun getNullable(): N? = null

  fun getDefaultNullability(): N = throw RuntimeException()
}

internal class NonNullableInsideNullable<T> {
  fun nonNullableTest(nonNullable: ParameterizedNonNullable<T & Any>) {}
}

interface Marker

open class Recursive<T> where T : Recursive<T>, T : Marker?

open class RecursiveNullable<T> where T : RecursiveNullable<T>?, T : Marker?

class RecursiveChild : Recursive<RecursiveChild>(), Marker

class RecursiveNullableChild : RecursiveNullable<RecursiveNullableChild?>(), Marker

class RecursiveParam<T> where T : Recursive<T>, T : Marker?

class RecursiveNullableParam<T> where T : RecursiveNullable<T>?, T : Marker?

fun <T> testRecursive() where T : Recursive<T>, T : Marker? {
  val generic = RecursiveParam<T>()
  val parametrized = RecursiveParam<RecursiveChild>()
}

fun <T> testRecursiveNullable() where T : RecursiveNullable<T>?, T : Marker? {
  val generic = RecursiveNullableParam<T>()
  val parametrized = RecursiveNullableParam<RecursiveNullableChild?>()
}
